### PR TITLE
HAL: Remove analogPinToDigitalPin

### PR DESCRIPTION
**Pull request by WickedShell**
`analogPinToDigitalPin(uint8_t)` was an old function used with the old AVR stuff. On all the interesting boards at the moment it generally returns a `-1` or does a direct pass through of the pin. This was only implemented in F4Light and PX4 builds. This was needed for a rover feature (something about auto trigger pins), and camera feedback setup.

The main note about this is that camera and rover can now ask you to check an arbitrary pin that the user has specified. I came across a UB in GPIO_RPI with reading where the pin number was used as the shift width. The last commit fixes the read to always be (silently) safe, however this behaviour seems to be much more spread through the GPIO code (particularly in GPIO_RPI). @lucasdemarchi I can push up checks on `GPIO_RPI::write()` that pin is less then 32 bits as well, but soome of the other macros use div/mod 10 so I'm a bit unsure what the GPIO range should be constrained to (and the fact that large pins just overflow is surprising). I'm only worried about the UB because users can specify some of the pins as parameters. (The write is particularly bad, as the relay pins documentation include values for the UI that would be out of range). This is an exsisting bug though, so if you aren't happy with this solution/want to look at it separately @lucasdemarchi I can drop that commit for now.

### CODE REVIEW

File: libraries/AP_HAL_Linux/GPIO_RPI.cpp [hunk 1]
```diff
@@ -29,7 +29,7 @@
 #define GPIO_MODE_ALT(g,a)  *(_gpio+(((g)/10))) |= (((a)<=3?(a)+4:(a)==4?3:2)<<(((g)%10)*3))
 #define GPIO_SET_HIGH       *(_gpio+7)  // sets   bits which are 1
 #define GPIO_SET_LOW        *(_gpio+10) // clears bits which are 1
-#define GPIO_GET(g)         (*(_gpio+13)&(1<<g)) // 0 if LOW, (1<<g) if HIGH
+#define GPIO_GET(g)         ((g >= 32) ? 0 : (*(_gpio+13)&(1<<g))) // 0 if LOW, (1<<g) if HIGH
```
Comment (lucasdemarchi) [inline libraries/AP_HAL_Linux/GPIO_RPI.cpp]: IMO the check should be on the caller, not here.

**Comment by lucasdemarchi** [comment]
> I'm only worried about the UB because users can specify some of the pins as parameters. (The write is particularly bad, as the relay pins documentation include values for the UI that would be out of range). This is an exsisting bug though, so if you aren't happy with this solution/want to look at it separately

yep, this is a separate bug/discussion. Here we can just put the protection on the HAL's gpio code so we don't do bad things as overriding the nearby peripheral registers (or crashing if we are in the end of the table).

Changing the pin number via parameter on Linux boards should never be done unless people really know what they are doing.

File: libraries/AP_HAL_Linux/GPIO_RPI.cpp (hunk 1 above)
Comment (WickedShell) [reply libraries/AP_HAL_Linux/GPIO_RPI.cpp]: Moved. I had stuck it in the macro to ensure any future users of the macro were protected, but it's currently only used in one location anyways so was easy to put in the caller.

**Comment by WickedShell** [comment]
> Changing the pin number via parameter on Linux boards should never be done unless people really know what they are doing.

Are there not a set of GPIO pins broken out for usage as relays/GPIO on most of the Linux boards?

**Comment by lucasdemarchi** [comment]
> Are there not a set of GPIO pins broken out for usage as relays/GPIO on most of the Linux boards?

There are the board gpios, if they are broken out for external usage is very board-dependent. I guess there are not many users with Linux boards using it or  they fall in the category "they really know what they are doing" that I mentioned above.

IMO the interface presented to the GCS should change. Listing in the doc the *physical* pin for each board does not scaled. We should rather leave it as `0 - MAX` or `1 - MAX` and let each board map to its physical pin.

Since you are mentioning rpi-based boards, these are the physical pins:
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_HAL_Linux/GPIO_RPI.h

**Pull request merged**